### ISSUE TITLE
Union resolution fail with `Unknown` elements

### ISSUE THREAD

**Issue by minmax** [microsoft/pyright#8449]
After updating pyright 1.1.371 -> 1.1.372, union resolution start to failed in some strange case.

```python
from __future__ import annotations
from collections.abc import Callable
from typing import Generic, TypeVar

T_co = TypeVar("T_co", covariant=True)
E_co = TypeVar("E_co", covariant=True)
F = TypeVar("F")


class Ok(Generic[T_co]):
    def or_else(self, op: object) -> Ok[T_co]: ...

class Err(Generic[E_co]):
    def or_else(self, op: Callable[[E_co], Result[T_co, F]]) -> Result[T_co, F]: ...

Result = Ok[T_co] | Err[E_co]


def inner(func: Callable[[E_co], Err[F]], r: Result[T_co, E_co]) -> Result[T_co, F]:
    match r:
        case Ok():
            return r.or_else(func)
        case Err():
            return r.or_else(func)  # Return type, "Ok[Unknown] | Err[F@inner]", is partially unknown
```

I found how to fix this, but I'm not sure if the error behavior is correct.
So if i change `Callable[[E_co], Err[F]]` to `Callable[[E_co], Result[Never, F]]`, the error goes away.

```python
def inner(func: Callable[[E_co], Result[Never, F]], r: Result[T_co, E_co]) -> Result[T_co, F]:
```

ps in real code i use [result](https://github.com/rustedpy/result) lib with Ok, Err and Result.

**Comment by erictraut** [comment]
I added some additional heuristics that allow me to strike a balance between the old behavior and the new behavior. This allows your use case to work while also addressing the other bugs that prompted this change. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.373.
